ruby/bps-kafka/lib/bps/kafka.rb

```diff
@@ -4,34 +4,52 @@
 module BPS
   module Kafka
     # ReliablePublisher is a slow, but reliable Kafka Publisher.
-    # There is no batching at all, messages are emitted synchronously.
+    # There is no batching at all, messages are delivered one by one.
     class ReliablePublisher < BPS::Publisher::Abstract
-      DELIVER_MESSAGE_RETRIES = 3 # 1 by kafka-ruby's Kafka::Client default, increase slightly
-
       class Topic < BPS::Publisher::Topic::Abstract
-        def initialize(client, topic_name)
-          @client = client
-          @topic_name = topic_name
+        def initialize(producer, topic_name, opts = {})
+          @producer = producer
+          @opts = opts.merge(topic: topic_name)
         end
 
         def publish(msg_data)
-          @client.deliver_message(msg_data, topic: @topic_name, retries: DELIVER_MESSAGE_RETRIES)
+          @producer.produce(msg_data, **@opts)
+          @producer.deliver_messages
           nil
         end
       end
+      private_constant :Topic
+
+      def initialize(seed_brokers, opts = {})
+        @producer_opts = extract_prefixed_opts!(opts, 'producer_')
+        @produce_opts = extract_prefixed_opts!(opts, 'produce_')
 
-      def initialize(seed_brokers, client_id: nil)
-        @client = ::Kafka.new(seed_brokers, client_id: client_id)
+        @client = ::Kafka.new(seed_brokers, **opts)
       end
 
       def topic(name)
-        Topic.new(@client, name)
+        producer = @client.producer(**@producer_opts)
```

Comment:
does the producer need to be closed explicitly? if so I suggest we memoize our topics

Reply:
@dim I won't bother with thread-safety (mutex) for memo for now. Or should I?